grin/src/Transformations/ExtendedSyntax/Conversion.hs

```diff
@@ -132,35 +153,62 @@ instance Convertible Exp New.Exp where
        of Binding Pattern Simplification to a more concise form.
 
       v.0 <- pure <value>
-      <non-var pat> <- pure v.0
+      <node pat> <- pure v.0
       <rhs2>
 
-      <non-var pat> @ v.0 <- pure <value>
+      <node pat> @ v.0 <- pure <value>
       <rhs2>
     -}
     (EBind lhs1 (Var var) rhs1)
-      | EBind (SReturn (Var var')) pat rhs2 <- rhs1
-      , isn't _Var pat
+      | EBind (SReturn (Var var')) (ConstTagNode tag args) rhs2 <- rhs1
+      , var == var'
+      -> do
+         newNodePat <- oldNodePatToAsPat tag args var
+         pure $ New.EBindF lhs1 newNodePat rhs2
+    {- NOTE: The following transformation can be done, because
+       unit and literal patterns are redundant. If the variable has
+       the same value as the pattern, then we can safely remove the
+       binding. If the variable holds some value different from the pattern,
+       then the program's behaviour is undefined, so we can do anything
+       with it.
+
+
+      v.0 <- pure <value>
+      <pat> <- pure v.0     -- pat is neither a node pat nor a var pat
+      <rhs2>
+
+      v.0 <- pure <value>
+      <rhs2>
+    -}
+      | EBind (SReturn (Var var')) valPat rhs2 <- rhs1
+      , isn't _Var valPat
       , var == var'
-      -> pure $ New.EBindF lhs1 (New.AsPat (convert var) (convert pat)) rhs2
+      -> pure $ New.EBindF lhs1 (New.VarPat $ convert var) rhs2
     {- NOTE: In this case, v.0 has been defined earlier in the program.
        This is a more general case that covers the one before as well.
 
       v.0 <- pure <value>
       <...>
-      <non-var pat> <- pure v.0
+      <node pat> <- pure v.0
       <rhs>
 
       v.0 <- pure <value>
       <...>
-      <non-var pat> @ a.0 <- pure v.0
+      <node pat> @ a.0 <- pure v.0    -- a.0 is a fresh variable
       <rhs>
     -}
-    (EBind lhs pat rhs) | isn't _Var pat -> do
-      asPatName <- deriveNewName "a"
-      pure $ New.EBindF lhs (New.AsPat (convert asPatName) (convert pat)) rhs
+    (EBind lhs (ConstTagNode tag args) rhs) -> do
+      asPatName  <- deriveNewName "a"
```

Comment:
Nit: I would prefer some better name for the derived name variables to show when the new name is introduced. Like "conv"

Reply:
Good point.
```suggestion
      asPatName  <- deriveNewName "conv"
```